src/charm.py

```diff
@@ -4,110 +4,123 @@
 
-    @property
-    def app_peer_data(self):
-        """Application peer data object."""
-        if not self.peers:
-            return {}
+        self.database_provides = relations.database_provides.Relation(
+            data_interfaces.DatabaseProvides(self, relation_name=DATABASE_PROVIDES_RELATION)
+        )
+        self.framework.observe(
+            self.database_provides.interface.on.database_requested,
+            self._reconcile_database_relations,
+        )
+        self.framework.observe(
+            self.on[DATABASE_PROVIDES_RELATION].relation_broken, self._reconcile_database_relations
+        )
 
-        return self.peers.data[self.app]
+        self.framework.observe(
+            getattr(self.on, "mysql_router_pebble_ready"), self._on_mysql_router_pebble_ready
+        )
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.leader_elected, self._on_leader_elected)
 
-    @property
-    def unit_peer_data(self):
-        """Unit peer data object."""
-        if not self.peers:
-            return {}
+        self.workload = workload.Workload(self.unit.get_container(MYSQL_ROUTER_CONTAINER_NAME))
 
-        return self.peers.data[self.unit]
+        self.tls = relations.tls.MySQLRouterTLS(self)
 
     @property
-    def endpoint(self):
+    def _endpoint(self) -> str:
         """The k8s endpoint for the charm."""
         return f"{self.model.app.name}.{self.model.name}.svc.cluster.local"
 
-    @property
-    def unit_hostname(self) -> str:
-        """Get the hostname.localdomain for a unit.
+    def _determine_status(self, event) -> ops.model.StatusBase:
+        inactive_relations = []
+        for relation, active in [
+            (DATABASE_REQUIRES_RELATION, self.database_requires.is_desired_active(event)),
+            (DATABASE_PROVIDES_RELATION, self.database_provides.is_desired_active(event)),
+        ]:
+            if not active:
+                inactive_relations.append(relation)
+        if inactive_relations:
+            return ops.model.WaitingStatus(
+                f"Waiting for relation{'s' if len(inactive_relations) > 1 else ''}: {', '.join(inactive_relations)}"
+            )
+        if not self.workload.container_ready:
+            return ops.model.MaintenanceStatus("Waiting for container")  # TODO
+        return ops.model.ActiveStatus()
 
-        Translate juju unit name to hostname.localdomain, necessary
-        for correct name resolution under k8s.
+    def _set_status(self, event=None) -> None:
+        if isinstance(self.unit.status, ops.model.BlockedStatus):
+            return
+        self.unit.status = self._determine_status(event)
 
-        Returns:
-            A string representing the hostname.localdomain of the unit.
-        """
-        return f"{self.unit.name.replace('/', '-')}.{self.app.name}-endpoints"
+    def _create_user(self) -> None:
+        if self.database_provides.active:
```

Comment:
AFAIK, this will prevent a 2nd user being created for a second related application.
Currently, the metadata allows multiple app relation (no `limit=1`) and this is one of the refactor desired goals